Super-streaming SXSSF

mobreza opened the pull request:
https://bz.apache.org/bugzilla/show_bug.cgi?id=63100

SXSSF works as designed and manages a small memory footprint when generating large files from a database.  But it only writes data to an output stream once everything has been written to SXSSF. This is problematic when used in web applications:

In our use case (our website allows users to generate Excel from the database), generating the SXSSF on the server takes about 5 minutes. Most clients give up within a minute (or the browser does it automatically), or the proxy times out due to no data being sent. Some users also retry the download request. A new request for download is initiated (while the server is busy generating the SXSSF for a client that already gave up). This can potentially lead to DOS.

To work around this issue, I've implemented a super-streaming version of SXSSF, a `SuperSXSSF`, that relies on `rowWriter` callback to generate row data.

With this approach our service is able to stream the generated Excel directly to the client and, best of all, is terminated in case the user cancels the download request. 

The `SuperSXSSF` prevents both download timeouts and potential DOS, while allowing developers all other XSSF actions (i.e. define styles) that don't take much processing time.

pjfanning committed:
lgtm warnings

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1841984 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
fix javadoc

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1841987 13f79535-47bb-0310-9956-ffa450edef68

Alain-Bearez committed:
adapting some contributions by Axel Richter on SO

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1841988 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
[bug-62736] Relations on XSLFPictureShape are removed unconditionally. Thanks to Mate Borcsok

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842055 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60656 - Support export file that contains emf and render it correctly

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1842056 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60656 - Support export file that contains emf and render it correctly

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1842059 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
[bug-62768] synchronize OPCPackage#close()

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842142 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
make OPCPackage#close() unsynchronized and add javadoc note saying method is not thread-safe

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842171 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
fix lgtm issues

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842216 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
fix gradle build issue

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842217 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
fix lgtm issues

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842218 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Remove outdated comment for japicmp

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842274 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Jenkins DSL: Remove JDK 9 and adjust location of image for the project-logo

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842275 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
lgtm issues

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842302 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
use StaxHelper to get event factory

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842402 13f79535-47bb-0310-9956-ffa450edef68

Alain-Bearez committed:
extract valuable code from 54470 submitted patch

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842548 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
try to set up xmlbeans build

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842599 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
try to fix xmlbeans jenkins job

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842629 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
[bug-62796] remove use of Stax Namespace class from PackagePropertiesMarshaller

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842647 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
rename xmlbeans job

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842663 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
get xmlbeans job to build dist artifacts

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842666 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
get xmlbeans job to build dist artifacts

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842667 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
disable junit xml collection

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842668 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
change method to get XMLEventFactory due to issues seen on Android SDK

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842683 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
deprecate unused Nullable class

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842684 13f79535-47bb-0310-9956-ffa450edef68

Alain-Bearez committed:
change chart data to experiment diverse scripts

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842687 13f79535-47bb-0310-9956-ffa450edef68

Alain-Bearez committed:
fix lgtm alerts in examples

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842707 13f79535-47bb-0310-9956-ffa450edef68

Alain-Bearez committed:
XDDF improvements based on StackOverflow examples

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842728 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
[bug-62800] Fix null pointer exception if a picture shape has no blip id

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842782 13f79535-47bb-0310-9956-ffa450edef68

Alain-Bearez committed:
more XDDF improvements based on SO examples

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842876 13f79535-47bb-0310-9956-ffa450edef68

Alain-Bearez committed:
example from StackOverflow question 47065690

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842925 13f79535-47bb-0310-9956-ffa450edef68

Alain-Bearez committed:
adding tricks from other answers on StackOverflow

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842959 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
use jaxp transformer instead of dom serializer to try fix old-xerces test failure

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842974 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
use jaxp transformer instead of dom serializer to try fix old-xerces test failure

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842979 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
small adjustments to byte sizes in TestPackage zip-bomb tests to try fix test that fails on old-xerces harness

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1842980 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
add stax entries to forbidden apis due to android issues

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1843009 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60656 - Support export file that contains emf and render it correctly

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1843025 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
merge trunk

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1843032 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
[bug-62807] fix typo in XSLFTableCell#removeBorder

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1843180 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
Bug 60713 - (S)XSSFWorkbook/POIXMLDocument.write(OutputStream) closes the OutputStream

regression - when reading the input from a zip stream

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1843341 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60656 - Support export file that contains emf and render it correctly

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1843342 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
add ignored resources

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1843354 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
merge trunk

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1843355 13f79535-47bb-0310-9956-ffa450edef68

Alain-Bearez committed:
fix most alerts reported by LGTM on OOXML

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1843481 13f79535-47bb-0310-9956-ffa450edef68

tballison committed:
62815 -- some numeric values not correctly extracted from xlsb

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1843553 13f79535-47bb-0310-9956-ffa450edef68

Alain-Bearez committed:
fix more LGTM alerts, in tests

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1843557 13f79535-47bb-0310-9956-ffa450edef68

Gagravarr committed:
Failing unit test for bug #62831

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844115 13f79535-47bb-0310-9956-ffa450edef68

Gagravarr committed:
#62831 Fix WorkbookFactory.create with a subclass of File, eg from JFileChooser

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844116 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Bug 62373: Support for FREQUENCY function

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844238 13f79535-47bb-0310-9956-ffa450edef68

Commit:
#62834 FormulaEvaluator.evaluateInCell() throws Exception

added cell type = formula check when looping through the shared formula range, to ignore any non-formula cells.

Also refactored a bit to enable passing in the evaluation context, as getCellFormula() uses it behind the scenes when evaluating a shared formula cell (has to shift the formula references based on the master cell).  Review of these changes is welcome, as always.

Checked all other code referencing the "SHARED" enum, and didn't see anything else that dealt with formula cell values and thus would need to notice non-formula cells.

Added unit test based on the failing file from Bugzilla.  Test failed until the fixed code was in place.

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844295 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
fix class cast issur recently introduced in BaseXSSFFormulaEvaluator

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844311 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60656 - Support export file that contains emf and render it correctly

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1844380 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60656 - Support export file that contains emf and render it correctly

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1844522 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60656 - Support export file that contains emf and render it correctly

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1844524 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Remove the workaround for a bug in JDK 11 which was fixed in ea+28

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844874 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Jenkins DSL: Adjust the view-description, but it is still commented out

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844875 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Set NOSONAR and reformat code

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844876 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Ignore directory ooxml-testlib

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844877 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Set NOSONAR, adjust test slightly and reformat code

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844878 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Typos and IDE warnings

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844879 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Adjust sample for creating comments to also create a .xlsx file
Enhance workbook factory to allow to create new empty workbooks as well

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844881 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
IDE warnings, slightly more tests and fix test to not leave a modified file behind

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844894 13f79535-47bb-0310-9956-ffa450edef68

tballison committed:
bug 62859 -- fix two potential NPEs when initializing XWPFSDTContent

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844920 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60656 - Support export file that contains emf and render it correctly

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1844931 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
update xmlbeans to 3.0.2

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1844955 13f79535-47bb-0310-9956-ffa450edef68

tballison committed:
bug 62624 -- fix npe by adding processing of module name mapping in VBAMacroReader

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845138 13f79535-47bb-0310-9956-ffa450edef68

Gagravarr committed:
Update for latest/dev javadocs at /apidocs/dev/

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845222 13f79535-47bb-0310-9956-ffa450edef68

Gagravarr committed:
More javadoc updates

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845227 13f79535-47bb-0310-9956-ffa450edef68

tballison committed:
bug 62625 -- add special handling for REFERENCE_NAME record that may only
contain an ascii string, against the spec in VBAMacroReader

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845238 13f79535-47bb-0310-9956-ffa450edef68

tballison committed:
bug 62625 -- clean up imports

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845246 13f79535-47bb-0310-9956-ffa450edef68

Gagravarr committed:
List additional dependencies from the Ant build file in Eclipse

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845255 13f79535-47bb-0310-9956-ffa450edef68

tballison committed:
bug 62624 -- fix recursion

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845284 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60656 - Support export file that contains emf and render it correctly

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1845291 13f79535-47bb-0310-9956-ffa450edef68

tballison committed:
bug 62624 -- fix loop identified as dodgy by FindBugs; add a other sanity checks.

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845299 13f79535-47bb-0310-9956-ffa450edef68

tballison committed:
bug 62624 -- ensure streams are closed...thanks to LGTM, and PJ!

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845300 13f79535-47bb-0310-9956-ffa450edef68

Gagravarr committed:
Allow the forrest task to be called from the site sub-project

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845343 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
IDE warnings, tried to reproduce Bug 58927, but could not

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845434 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Remove JavaDoc warnings

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845435 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60656 - Support export file that contains emf and render it correctly

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1845496 13f79535-47bb-0310-9956-ffa450edef68

tballison committed:
bug 62624 -- further cleanup

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845511 13f79535-47bb-0310-9956-ffa450edef68

tballison committed:
bug 60316 -- until we can implement it properly, gracefully skip the glossary document in XWPF.

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845517 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Bug 62836: Implementation of Excel TREND function

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845586 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60656 - Support export file that contains emf and render it correctly

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1845612 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
merge trunk

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1845617 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
Bug 62872 - Writing large files with 800k rows gives java.io.IOException: This archive contains unclosed entries.

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1845629 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60656 - Support export file that contains emf and render it correctly

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1846472 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
add testcase for https://bz.apache.org/bugzilla/show_bug.cgi?id=62906

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846489 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
add testcase for https://bz.apache.org/bugzilla/show_bug.cgi?id=62906

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846492 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
fix typo

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846497 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
[bug-62908] populate NamedColors

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846520 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
remove unused import

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846548 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
remove duplicate attempt to add column headers

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846549 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
put back XSSFColor(java.awt.Color clr) constructor

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846701 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Try to create the temporary directory for ImageIO to see if that makes 
the tests work again when built with Maven

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846706 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
One more try to avoid the test-failures related to ImageIO-cache dir

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846719 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
Try to fix ImageIO cache error

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846723 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
exclude H35 because of strange file permission errors

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846744 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
exclude H35 because of strange file permission errors

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846746 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
Multiproject test including Windows slaves

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846773 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62921 - Provide OOXMLLite alternative for Java 12+

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846809 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62921 - Provide OOXMLLite alternative for Java 12+

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846810 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62921 - Provide OOXMLLite alternative for Java 12+

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846811 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62921 - Provide OOXMLLite alternative for Java 12+

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846824 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62921 - Provide OOXMLLite alternative for Java 12+

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846849 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62921 - Provide OOXMLLite alternative for Java 12+

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846870 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
Remove/Minimize references to Scratchpad module from other modules

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846931 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
Try to workaround/ignore kept open files of JaCoCo ant task

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846945 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
Remove commented out System.out calls / Move DummyGraphics2d to testcase area, as it's not used by production code

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846946 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
release prepare for 4.0.1 - updating build.xml and changes.xml

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846949 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
prepare for 4.0.2

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1846951 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
[bug-62929] add null check for blip fill. Thanks to Mate Borcsok

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847004 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
tidy up code formatting

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847012 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
Remove duplicated .jar extension when crafting the maven-jar name

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847068 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
merge trunk

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1847069 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
add missing classes to poi-ooxml-schemas

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847083 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60656 - Support export file that contains emf and render it correctly

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1847209 13f79535-47bb-0310-9956-ffa450edef68

tballison committed:
bug 60316 -- skip the glossary document, not the template document.

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847263 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
fixed escaping of sheet names. Thanks to gallonfizik. This closes #134

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847418 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62949 - SlideShow rendering - keyframe fractions must be increasing

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847428 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62951 - FileMagic not correctly identified

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847429 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Exclude ooxml-testlib from src-zip/tar.gz

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847436 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Add test for FileMagic
JavaDoc warning fixes
Remove some IDE warnings
Reformat code of sample application

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847437 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60713 - (S)XSSFWorkbook/POIXMLDocument.write(OutputStream) closes the OutputStream

regression

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847438 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
Fix exception for old powerpoint documents without the usual stream name

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847442 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Fix TestFileMagic

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847454 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
Fix exception for old powerpoint documents without the usual stream name

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847487 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
exclude sha256 for nexus upload

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847493 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
release prepare for 4.0.1 - updating build.xml and changes.xml

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847494 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
prepare for 4.0.2

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847496 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
update links to POI 4.0.1

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1847861 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
add 4.0.1 to the doap file

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848115 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
[bug-57342] add SXSSFWorkbook setZip64Mode method

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848179 13f79535-47bb-0310-9956-ffa450edef68

Alain-Bearez committed:
create chart from scratch by sandeeptiwari32

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848432 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
lgtm issues

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848438 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62953 - Rendering of FreeformShapes with formula fails

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848492 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62966 - init presetShapeDefinitions.xml fail under IBM jdk

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848494 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62994 - IBM JCE workarounds

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848538 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62994 - IBM JCE workarounds
add policy restriction exemption

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848539 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
[bug-62996]try suggested test change

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848591 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62999 - IBM JDK JIT causes AIOOBE in TexturePaintContext

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848601 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62999 - IBM JDK JIT causes AIOOBE in TexturePaintContext
refactor jacoco/junit calls and add more IBM JIT excludes

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848642 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
add ooxml lite agent again ...

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848655 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
disable findbugs for IBM JDK

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848809 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
disable findbugs for IBM JDK

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848811 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
merge trunk

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1848906 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#63011 - Multiple digital signature in excel file broke first signature

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848974 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
ignore keystore files on integration test

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1848975 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62365 - SVG image support in XSLF

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849030 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
activate IBMJDK again, unlimited JCE policy files are now included by default, and the other IBMJDK issues (#62994,#62966,#62999) have been fixed

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849032 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
merge trunk

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1849035 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
merge trunk

git-svn-id: https://svn.apache.org/repos/asf/poi/branches/hemf@1849036 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#60656 -  EMF image support in slideshows

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849040 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
fix constructor after its delegated now

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849041 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
Implemented GEOMEAN function. Thanks to gallonfizik. This closes #136

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849042 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
add missing license

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849043 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
fix findbugs issue

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849044 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62365 - SVG image support in XSLF

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849046 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
fix junit tests

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849047 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
use spaces for indents

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849067 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
add null check in XLSFPictureShape

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849069 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
changed to bouncer pattern - to match also the rest of the code
see https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849074 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
[bug-63005] disable the reading of DTDs altogether

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849123 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
lgtm issues: potential index out of bounds exceptions

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849127 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
[bug-63005] ignore openxml4j/ContentTypeHasEntities.ooxml

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849128 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
only refactorings to break up the long method

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849130 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
[bug-63013] add XWPFRun setLang method

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849152 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
[github-137] solves/unifies blank/missing value colection for PRODUCT/MDETERM/GEOMEAN. Thanks to gallonfizik. This closes #137

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849237 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#63017 - Remove rows from a XSLFTable

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849244 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#63017 - Remove rows from a XSLFTable

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849248 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
fix index check in HemfDraw

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849250 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#62886 - Regression extracting text from corrupted docx files

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849252 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Add missing jacoco-lib to test-scratchpad
Remove some xmlns that are not needed any more now

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849280 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
One more file-leak-detector exclude

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849286 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Rework test somewhat and enable logger to better indicate if the test is
failing because JDK options are missing with JDK >= 9

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849287 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
fix old xerces errors because of not available disallow-doctype parser feature

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849336 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Add more file-leak exlcudes for the mass regression testing

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849715 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Code-style: Unify how arrays are specified from C-style to normal Java-style

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849716 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Bug 62884: sheetnum is not checked in InternalWorkbook.setSheetHidden()

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849717 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
JavaDoc fixes/additions and some IDE warnings removed

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849718 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Simplify some string operations

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849760 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Remove unnecesssary semicolons

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849761 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Fix incorrect constant for pict-files

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849762 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Fix various IDE warnings and some JavaDoc adjustments

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849763 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Bug 62839: Adjust MathX.floor() to handle negative values the same
way as Excel/LibreOffice

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849764 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Add some simple coverage of HSLF-Dev-Tools

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849765 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
IDE and JavaDoc fixes, close() in tests

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849812 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Re-instate two dev-tools that we lost in some refactoring,

at least OOXMLPrettyPrint is still very useful for comparing ooxml-files

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849813 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Bug 62629: Allow to handle files with invalid content types for pictures

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849814 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Adjust/enhance tests for bug 62275 some more

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849836 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Adjust test to not require src/documentation to be available

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849837 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Add XMLBeans jobs for JDK 8 and 11.
Add initial version of Jenkins DSL for the Github-PullRequest build-job

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849838 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Add test which verifies that bug 61605 is already fixed

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849878 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Use normal string parameter for now

Plugin for gitParam is not installed in the Apache Jenkins instance

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849879 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Bug 61532: Fix setting values/types during formula evaluation for SXSSF

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849880 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Run Cell-tests on HSSF as well and make handling of null-values consistent across implementations

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849881 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Jenkins DSL: throttle Github PR builds to not run many at once

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849884 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Jenkins DSL: enable Github PR builds, it seems to be functional now

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849885 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Jenkins DSL for Github PR build: Try to create the one file which is read from src/documentation during target 'jenkins'

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849887 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
Zero out garbage bytes in FontEntityAtom

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849895 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
Fix Javadocs

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849896 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
Add Enum method to help identifying unknown records with POI-Visualizer

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849897 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#63028 - Provide font embedding for slideshows

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849898 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
IDE and JavaDoc fixes, make base test classes abstract

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849968 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Bug 60845: Apply patch and adjust tests

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849969 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Jenkins DSL: Disable XMLBeans-JDK-11 as it will not work yet

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1849970 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
IDE and JavaDoc fixes

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850007 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Bug 60460: Handle null workbook or sheet names and emit #REF as Excel does
instead of throwing NullPointerException

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850008 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Adjust test for bug 60460 to also run fine on Windows

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850022 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
#63047 - Make POILogger subclassable

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850040 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
exclude integrationtest/build from rat-check

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850041 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
fix since annotation

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850048 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Add missing file-close in ZipInputStreamZipEntrySource seen in some mass regression tests

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850114 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Add missing revert() in OPCPackage to close Zip-streams always

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850115 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Bug 62307: made Cell#getNumericCellValue() behavior consistent across HSSF/XSSF/SXSSF.\nAll three implementations throw ISE when trying to get numeric value from a boolean-valued cell, have it a formula set or not.

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850207 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Bug 62828: CellReference(Cell) now initializes sheet name.
Changed CellReference to CellAddress in XSSFHyperlink because it is what it should return.
Updated all relevant tests, added a test for CellReference(Cell).

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850210 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Bug 62993: XSSFEvaluationSheet now retrieves valid last row index from underlying XSSFSheet. Thanks to Axel Howind.

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850212 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Improved test coverage for *Cell classes

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850338 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Bug 63057: made SXSSFCell.setCellValue(String|RichTextString) exception-safe.
Also removed unused code.

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850342 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Fixed build (explicitly specified charset for new String(byte[]))

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850366 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
throw IllegalArgumnetException if null directory specified

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850394 13f79535-47bb-0310-9956-ffa450edef68

Commit:
updated mockito to 2.23.4 to fix java 11 build

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850590 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Bug 62904: Support array arguments in IF and logical IS*** functions

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850646 13f79535-47bb-0310-9956-ffa450edef68

Commit:
unified setCellType(null/_NONE) logic, setCellFormula(null) logic. updated javadoc and tests

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850676 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Bug 63054: improved evaluation of array formulas with error in arguments

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850742 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Bug 63054: updated formulas in test spreadsheet to keep build happy

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850772 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
use spaces for indents

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1850926 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Bug #60977 writing XSSF with custom properties twice corrupts output

turns out the output XML data buffer for the custom properties document was appended to, not replaced, when writing in the presence of existing content.  Now clearing output buffer first in this case.  When first creating custom properties, i.e. first write after changes, there is no buffer yet, so nothing to clear.

Does not affect any read operations, or other writes beyond custom properties, as that was where the problem lay, not down in the Zip stuff.

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851084 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Add one more test for bug 62828 and fix some IDE warnings

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851088 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
IDE and JavaDoc fixes, move/remove some unfixed tests

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851089 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Enable weekly run of Java 12 build

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851090 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Remove obsolete settings from JDK 11/12 build

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851091 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Revert inadvertently commented lines

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851092 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Add reproducer for bug 60355

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851208 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Bug 62810: AreaReference ctor looses sheet name if rows or columns swapped

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851209 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Fix some Findbugs and IDE issues, refactor some duplicated code,
improve some exception texts, add comment for missing Ptg for SxName

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851210 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Don't fail on row-nums larger than Integer.MAX_VALUE, we had one sample
file which triggered this

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851211 13f79535-47bb-0310-9956-ffa450edef68

Commit:
follow-up to Bug 62904. More tests and improved evaluation of IF in array mode

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851263 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
[github-140] add .travis.yml. Thanks to krichter722. This closes #140

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851312 13f79535-47bb-0310-9956-ffa450edef68

kiwiwings committed:
update xmlbeans targets

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851657 13f79535-47bb-0310-9956-ffa450edef68

pjfanning committed:
try to fix xmlbeans build

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851667 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Fix some IDE warnings and JVM deprecations

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851795 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Fix passing in --add-opens for Java 9+ after the poiunit refactoring, this
only caused problems with Java 9+ on Windows, which we do not run in CI (yet)

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851796 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Fix invalid comment in build.xml

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851797 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Jenkins: Add initial build for JDK 13, add build for JDK 12 on Windows

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851798 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Jenkins: Try to disable coverage for JDK 13 for now as it seems to cause issues

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851803 13f79535-47bb-0310-9956-ffa450edef68

centic9 committed:
Add required --add-opens and other settings for JDK 12 on Windows

git-svn-id: https://svn.apache.org/repos/asf/poi/trunk@1851804 13f79535-47bb-0310-9956-ffa450edef68

mobreza committed:
Super-streaming SXSSF

asfgit commented:
Can one of the admins verify this patch?

mobreza commented:
The changes mostly make private vars protected for access in `SuperSXSSF`-* classes. Workbook is a copy of SXSSFWorkbook, but calls the `rowGenerator` before closing the stream.

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/streaming/IRowGenerator.java:
```diff
@@ -0,0 +1,32 @@
+	 * @param sheet the sheet
+	 * @throws Exception the exception
+	 */
+	void generateRows(SXSSFSheet sheet) throws Exception;
```
please use spaces for indents, not tabs

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/streaming/SXSSFSheet.java:
```diff
@@ -65,15 +65,23 @@ Licensed to the Apache Software Foundation (ASF) under one or more
-    private final TreeMap<Integer,SXSSFRow> _rows = new TreeMap<>();
-    private final SheetDataWriter _writer;
+    protected final SXSSFWorkbook _workbook;
+    private final TreeMap<Integer,SXSSFRow> _rows=new TreeMap<Integer,SXSSFRow>();
```
this line change is unnecessary

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/streaming/StreamingSheetWriter.java:
```diff
@@ -0,0 +1,87 @@
+    	throw new RuntimeException("Not supported with StreamingSheetWriter");
+    }
+    
+    @Override
```
no need for this finalize

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/streaming/StreamingSheetWriter.java:
```diff
@@ -0,0 +1,87 @@
+    
+    @Override
+    boolean dispose() throws IOException {
+    	try {
```
I don't think exception should be swallowed

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/streaming/StreamingSheetWriter.java:
```diff
@@ -0,0 +1,87 @@
+
+    @Override
+    public File createTempFile() throws IOException {
+    	throw new RuntimeException("Not supported with StreamingSheetWriter");
```
could you ensure consistent indents of 4 spaces per indent?

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/streaming/SheetDataWriter.java:
```diff
@@ -71,6 +71,12 @@ public SheetDataWriter() throws IOException {
         _out = createWriter(_fd);
     }
+    
+    // Constructor for StreamingSheetWriter
```
this comment has little value

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/streaming/StreamingSheetWriter.java:
```diff
@@ -0,0 +1,87 @@
+import org.apache.poi.util.POILogFactory;
+import org.apache.poi.util.POILogger;
+
+public class StreamingSheetWriter extends SheetDataWriter {
```
could you add a little javadoc with `@since 4.1.0` annotation?

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/streaming/SuperSXSSFSheet.java:
```diff
@@ -0,0 +1,60 @@
+
+import org.apache.poi.xssf.usermodel.XSSFSheet;
+
+public class SuperSXSSFSheet extends SXSSFSheet {
```
Can you add a little javadoc with `@since POI 4.1.0`? I'd prefer a more neutral descriptive class name.

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/streaming/SuperSXSSFWorkbook.java:
```diff
@@ -0,0 +1,346 @@
+import org.apache.poi.xssf.usermodel.XSSFSheet;
+import org.apache.poi.xssf.usermodel.XSSFWorkbook;
+
+public class SuperSXSSFWorkbook extends SXSSFWorkbook {
```
Can you add a little javadoc with `@since POI 4.1.0`? I'd prefer a more neutral descriptive class name.

pjfanning reviewed (changes requested):
Thanks. The changes look useful. My biggest concern is that there are no tests included. Could you add some?

mobreza commented:
@pjfanning I've cleaned up indentation and removed my notes (comments). I'm preparing a few unit tests.  Feel free to suggest a better name than `SuperSXSSF*`.

mobreza committed:
Copy data to output stream using `ISheetInjector`

- SXSSF and SuperSXSSF use different approaches
- Reduces code duplication in SuperSXSSFWorkbook
- Added unit tests

pjfanning commented:
Thanks. I'm struggling to think of ideal names for the new classes. Maybe you could put an `@Beta` annotation on the new classes and maybe just use a name like SXSSF2Workbook. I suspect that we will be able to merge SXSSF2Workbook and SXSSFWorkbook but we can do that later.

mobreza committed:
Added @Beta annotation

mobreza commented:
@pjfanning I've recycled existing SXSSFWorkbook tests. It was helpful in identifying a couple of bugs.

I wanted to also reduce the amount of duplicated code from SXSSFWorkbook and because `copyStreamAndInjectWorksheet` is static I opted for `ISheetInjector` approach.

pjfanning commented on src/ooxml/testcases/org/apache/poi/xssf/streaming/TestSuperSXSSFWorkbook.java:
```diff
@@ -0,0 +1,500 @@
+    	Row row = sheet.createRow(1);
+    	Cell cell = row.createCell(1);
+    	cell.setCellValue("value 2_1_1");
+    	SuperSXSSFWorkbook wb1 = new SuperSXSSFWorkbook(xssfWb1);
```
inconsistent indents in whole class

pjfanning commented on src/ooxml/testcases/org/apache/poi/xssf/streaming/TestSuperSXSSFWorkbook.java:
```diff
@@ -0,0 +1,500 @@
+
+    	SuperSXSSFWorkbook wb2 = new SuperSXSSFWorkbook(xssfWb2);
+    	assertEquals(1, wb2.getNumberOfSheets());
+    	Sheet sheet  = wb2.getStreamingSheetAt(0);
```
2 spaces before =

pjfanning commented on src/ooxml/java/org/apache/poi/xssf/streaming/IRowGenerator.java:
```diff
@@ -0,0 +1,35 @@
+     * @param sheet the sheet
+     * @throws Exception the exception
+     */
+    void generateRows(SXSSFSheet sheet) throws Exception;
```
I don't quite understand this - why not just add rows using Sheet#createRow?

https://github.com/apache/poi/pull/141/files#diff-ea5bbd436cd5c48efd0716b8d6f0e6ebR155 (highlight line in yellow) - seems a very verbose way to create a row

mobreza replied:
@pjfanning Row and cell generation happens in `SXSSFWorkbook#injectData()` and `#copyStreamAndInjectWorksheet()` and that allows for more direct streaming to the client. A call to `SuperSXSSFSheet#createRow` outside `generateRows()` is invalid as `SuperSXSSFSheet` does not have a `_writer` until `workbook.write(out);` is running.

https://github.com/apache/poi/pull/141/files#diff-ea5bbd436cd5c48efd0716b8d6f0e6ebR274 in `populateData()` demonstrates it better.

mobreza replied:
Not 100% true: you can call `#createRow` as long as you stay within `randomAccessWindowSize` of SXSSFSheet.

mobreza committed:
Code formatted

pjfanning commented:
xlsx files are zip files that contain XML files. There is an XML file for each sheet. There are also XML files for 'shared strings', styles and comments. How does this code base support these requirements?

mobreza commented:
@pjfanning SXSSF utilizes lots of XSSF, but avoids OutOfMemory exception by writing row data to a temporary file, the contents of which is injected at the right place when the final xlsx XML+ZIP is written to the output stream:

1. Prepare XSSF (styles, comments, sheet info....) in memory
1. Prepare entire SXSSF sheet XML content -- write row XML to a temporary file (this may take minutes)
1. Write XSSF to a new temporary xlsx file (contains everything but the row data from SXSSF)
1. Copy temporary xslx contents to output stream
1. ... inject row XML data from the temporary file (`SXSSFWorkbook#injectData()`)

Step 1 is quick. Most time is spent in step 2 to write XML row data to a temporary file. Steps 3 is quick. Steps 4 and 5 are quick, only copying existing XML from temporary files.

Writing the final xlsx to output stream begins in step 4, but is delayed (for minutes) until we've prepared the temporary file with all SXSSF row data.

SuperSXSSF postpones the generation of the sheet XML contents:

1. Prepare XSSF (styles, comments, sheet info....) in memory
1. Declare how SXSSF sheet XML content is generated using `setRowGenerator()`, but do not spend time generating row XML
1. Write XSSF to a new temporary xlsx file (contains everything but the row data from SXSSF)
1. Copy temporary ZIP+XML file contents to output stream
1. ... and generate and inject SXSSF XML row data on the fly (this may take minutes).

Steps 1 is quick. Step 2 is just a declaration and is executed quickly. Step 3 is quick. Step 4 can start copying XML from temporary files to the final output stream. Most time is spent in step 5 for generation of XML row data.

Writing the final xslx to output stream begins in step 4, but we have not spent any time generating SXSSF row data.

pjfanning commented:
Could you add a sample to src/examples to demo the usage?

mobreza commented:
@pjfanning Yes, I'll add some.

I've been thinking about `SXSSF2` that could be a decent hybrid of `SXSSF` and `SuperSXSSF`: keep the declarative nature of `SuperSXSSF` (postponing row generation until needed) and allow for all operations on `XSSFWorkbook` (styles, shared strings, etc.). A two-phase `XSSFWorkbook#write(out)` is then used:

1. write the `XSSFWorkbook` to temporary file
1. copy from temporary file to the final output stream the basic elements of the ZIP that are required for xlsx compatibility, but skip  styles, shared strings, etc.
1. write sheet data to the final output stream (including `SuperXSSF` delayed sheets (and here we could declare new styles, shared strings, etc.).
1. the `XSSFWorkbook` now contains updated styles, strings, etc. 
1. write the workbook with updates to another temporary file
1. copy the updated shared resources from this temporary file to the final output stream

This would allow us to start streaming xlsx immediately, make changes to common XMLs and still write out the full Excel file.

centic9 commented:
Unfortunately this was closed by some automated job, but we cannot reopen it ourselves, please reopen if this is still an active PR so it does not get lost.

jandhollander commented:
@mobreza @pjfanning Is this something that could make it in, in the future?
We have a web application that uses SXSSF and would like the download of a workbook to start immediately while generating the workbook.

pjfanning commented:
This PR has been corrupted but if someone wants to create a new one, we might consider it.

mobreza commented:
Is there a way to reopen this PR?

pjfanning commented:
@mobreza could you create a new MR from your branch?
